# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import numpy as np
import pandas as pd
import scipy.stats as sp
from scipy.stats import ttest_ind
from scipy.stats import ttest_rel
from scipy.stats import ttest_1samp

# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

Load the data from [here](https://drive.google.com/file/d/1CgaV3QwIw0vJEfANGX0pAjdwQ7mKqePd/view?usp=sharing) store them locally in the provided data folder and run the code below.

In [2]:
# Run this code to load the data

pokemon = pd.read_csv('../data/pokemon.csv')

Let's start off by looking at the `head` function in the cell below.

In [3]:
# Your code here:

pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [4]:
# Your code here:


In [5]:
Legendary=pokemon.loc[pokemon['Legendary'] == True]
Legendary['Legendary'].count()

65

In [6]:
NonLegendary=pokemon.loc[pokemon['Legendary'] == False]
NonLegendary['Legendary'].count()

735

There are 65 Legendary Pokemon vs 735 regular Pokemon

Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [7]:
# Your code here:
Legendary['Total'].mean()

637.3846153846154

In [8]:
Legendary['Total'].std()

60.93738905315346

In [9]:
NonLegendary['Total'].mean()

417.21360544217686

In [10]:
NonLegendary['Total'].std()

106.76041745713022

The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

In [32]:
# Your code here:

sp.stats.ttest_ind(Legendary.dropna()['Total'], NonLegendary.dropna()['Total'], equal_var=False)

Ttest_indResult(statistic=25.8335743895517, pvalue=9.357954335957446e-47)

In [34]:
#paolo:good approach-notice though that when you use
# dropna() as you do above you remove the entire row
# where there is a single NaN-> meaning that you are potentally
# discarding valid Total points.

What do you conclude from this test? Write your conclusions below.

# Your conclusions here:

pvalue is very small <0.05, We conclude that there is a significant difference between the mean of Legendary Pokemon and the mean of non legendary Pokemon 

How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [12]:
# Your code here:
Type1=pokemon[['Type 1','Total']]
Type1.groupby(['Type 1']).count()

Total
Type 1         
Bug          69
Dark         31
Dragon       32
Electric     44
Fairy        17
Fighting     27
Fire         52
Flying        4
Ghost        32
Grass        70
Ground       32
Ice          24
Normal       98
Poison       28
Psychic      57
Rock         44
Steel        27
Water       112

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [13]:
Type1.groupby(['Type 1']).std()

Total
Type 1              
Bug       117.875223
Dark      109.126217
Dragon    146.267538
Electric  105.721952
Fairy     123.781680
Fighting  102.464378
Fire      109.760496
Flying    161.400124
Ghost     110.072685
Grass     106.650626
Ground    123.913081
Ice       108.281027
Normal    115.733038
Poison     92.358687
Psychic   139.026645
Rock      108.059227
Steel     115.420907
Water     113.188266

In [14]:
Type1.groupby(['Type 1']).mean()

Total
Type 1              
Bug       378.927536
Dark      445.741935
Dragon    550.531250
Electric  443.409091
Fairy     413.176471
Fighting  416.444444
Fire      458.076923
Flying    485.000000
Ghost     439.562500
Grass     421.142857
Ground    437.500000
Ice       433.458333
Normal    401.683673
Poison    399.142857
Psychic   475.947368
Rock      453.750000
Steel     487.703704
Water     430.455357

In [15]:
# Your code here:
water=Type1.loc[Type1['Type 1'] == 'Water']

In [16]:
nonwater=Type1.loc[Type1['Type 1'] != 'Water']

In [17]:
water['Total'].mean()

430.45535714285717

In [18]:
water['Total'].std()

113.18826606431458

In [19]:
nonwater['Total'].mean()

435.85901162790697

In [20]:
nonwater['Total'].std()

121.09168230208066

Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

In [21]:
# Your code here:
sp.stats.ttest_ind(water.dropna()['Total'], nonwater.dropna()['Total'], equal_var=True)

Ttest_indResult(statistic=-0.4418547448849676, pvalue=0.6587140317488793)

Write your conclusion below.

# Your conclusions here:

pvalue is higher than 0.05, we conclude that there is not a significant difference between the mean of Water Pokemon and the mean of non water Pokemon 

# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

In [22]:
# Your code here:

#use paired t-test for dependent samples from the lecture
#use scipy function ttest_rel or ttest_1samp (1 sample test)

ttest_rel(pokemon['Attack'],pokemon['Defense'])

Ttest_relResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)

Describe the results of the test in the cell below.

# Your conclusions here:

pvalue is very small <0.05, we conclude that there could be a significant difference between the mean of Pokemon defense and attack scores.

We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

In [23]:
# Your code here:
ttest_rel(pokemon['Sp. Atk'],pokemon['Sp. Def'])


Ttest_relResult(statistic=0.853986188453353, pvalue=0.3933685997548122)

Describe the results of the test in the cell below.

# Your conclusions here:

pvalue is higher than 0.05, we conclude that there is not a significant difference between the mean of Pokemon special defense and the mean of Pokemon special attack

As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [35]:
# Your code here:

In [36]:
#create an extra column with values for the difference between Defense and Attack values
pokemon['Difference']=pokemon['Defense']-pokemon['Attack']

In [26]:
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  Difference  
0       65       65     45           1      False           0  
1       80       80     60           1      False           1  
2      100      100     80           1      False           1  
3      122      120     80           1      False          23  
4       60       50     65           1      False          -9

In [39]:
diff_mean=pokemon['Difference'].mean()

-5.15875

In [41]:
ttest_1samp(pokemon['Difference'], diff_mean)

Ttest_1sampResult(statistic=0.0, pvalue=1.0)

This does not look right to me, but I do not know what is the issue

In [42]:
#Paolo: the issue is that you should test the hyp that the diff is zero,
# not diff_mean- that is why you are getting such a weird result.
# (diff_mean is exactly the mean of pokemon['Difference']) so the test
# does not make much sense)
ttest_1samp(pokemon['Difference'], 0)

Ttest_1sampResult(statistic=-4.325566393330478, pvalue=1.7140303479358558e-05)

In [43]:
#paolo check  good work!

# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [29]:
# Your code here:



Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [30]:
# Your code here:



Based on a 95% confidence, should we reject the null hypothesis?

In [31]:
# Your answer here:

